utility/data_fetcher.go

```diff
@@ -0,0 +1,167 @@
+package utility
+
+import (
+	"errors"
+	"time"
+)
+
+var (
+	typeErr = errors.New("Invalid type for field.")
+	nilErr  = errors.New("Mandatory field missing.")
+)
+
+type DataFetcher struct {
+	Err error
+}
+
+func (d *DataFetcher) Float64(base map[string]interface{}, key string, keys ...string) float64 {
+	val := d.getDeep(base, keys...)[key]
+	if valFloat, ok := val.(float64); ok {
+		return valFloat
+	} else if val == nil {
+		d.Err = nilErr
+	} else {
+		d.Err = typeErr
+	}
+	return 0.0
+}
+
+func (d *DataFetcher) IntPtr(base map[string]interface{}, key string, keys ...string) *int {
+	val := d.getDeep(base, keys...)[key]
+	if val == nil {
+		return nil
+	} else if valFloat, ok := val.(float64); ok {
+		valInt := int(valFloat)
+		if valFloat == float64(valInt) {
+			return &valInt
+		}
+	} else if valFloat, ok := val.(float32); ok {
+		valInt := int(valFloat)
+		if valFloat == float32(valInt) {
+			return &valInt
+		}
+	} else if valInt, ok := val.(int); ok {
+		return &valInt
+	}
+	d.Err = typeErr
+	return nil
+}
+
+func (d *DataFetcher) Interface(base map[string]interface{}, key string, keys ...string) interface{} {
+	return d.getDeep(base, keys...)[key]
+}
+
+func (d *DataFetcher) Int(base map[string]interface{}, key string, keys ...string) int {
+	if val := d.IntPtr(base, key, keys...); val != nil {
+		return *val
+	}
+	val := d.getDeep(base, keys...)[key]
+	if val == nil {
+		d.Err = nilErr
+	}
+	return 0
+}
+
+func (d *DataFetcher) StringPtr(base map[string]interface{}, key string, keys ...string) *string {
+	val := d.getDeep(base, keys...)[key]
+	if val == nil {
+		return nil
+	} else if valStr, ok := val.(string); ok {
+		return &valStr
+	}
+	d.Err = typeErr
+	return nil
+}
+
+func (d *DataFetcher) String(base map[string]interface{}, key string, keys ...string) string {
+	if val := d.StringPtr(base, key, keys...); val != nil {
+		return *val
+	}
+	val := d.getDeep(base, keys...)[key]
+	if val == nil {
+		d.Err = nilErr
+	}
+	return ""
+}
+
+func (d *DataFetcher) StringArr(base map[string]interface{}, key string, keys ...string) []string {
+	val := d.getDeep(base, keys...)[key]
+	if val == nil {
+		return nil
+	} else if valArr, ok := val.([]string); ok {
```

Comment:
i think this case is not needed

Reply:
I'd prefer leaving it in as it would be really unexpected behavior when querying for a StringArray would not return anything when the value for the specified key is a `[]string`

Reply:
As reply for this and `Int` / `String` implementations above:
In apm-server data is meaningful - we know what is a transaction, error, span etc. 
Why have so much generic code as if we don't know what data is?

This case is never executed in production, yet you took the time to write it and test, me to read it, the compiler to compile it, etc...